### PR TITLE
[HPO] updates HPO integration

### PR DESCRIPTION
**Pull request by yunchu**
updated for HPO

- to apply hpopt updates, some required part in the OTE was updated
- enabled hpo to new tasks
    * object counting
    * rotated boundbox detection
    * anomaly tasks

### CODE REVIEW

**Comment by yunchu** [comment]
run ote-test

**Comment by yunchu** [comment]
run ote-test

**Comment by yunchu** [comment]
run ote-test

**Comment by yunchu** [comment]
run ote-test

**Comment by yunchu** [comment]
run ote-test

**Comment by yunchu** [comment]
run ote-test

**Comment by yunchu** [comment]
test completed with some failures but those are not related with this PR.
http://validationreports.sclab.intel.com:8006/reports/build_number_report?test_session_build_number=pr-ote-test-413&environment=iotg-dev-workstation-18

File: external/mmdetection/detection_tasks/extension/utils/hooks.py [hunk 1]
```diff
@@ -176,6 +178,8 @@ def before_iter(self, runner: BaseRunner):
 
     @check_input_parameters_type()
     def after_iter(self, runner: BaseRunner):
+        # put some runner's training status to use on the other hooks
+        runner.log_buffer.output['current_iters'] = runner.iter
```
Comment (eunwoosh) [inline external/mmdetection/detection_tasks/extension/utils/hooks.py:182]: Doesn't this code need to be only in after_epoch function?

File: external/mmsegmentation/segmentation_tasks/extension/utils/hooks.py [hunk 2]
```diff
@@ -173,6 +175,8 @@ def before_iter(self, runner: BaseRunner):
 
     @check_input_parameters_type()
     def after_iter(self, runner: BaseRunner):
+        # put some runner's training status to use on the other hooks
+        runner.log_buffer.output['current_iters'] = runner.iter
```
Comment (eunwoosh) [inline external/mmsegmentation/segmentation_tasks/extension/utils/hooks.py:179]: It's the same question as above, doesn't this code need to be only in after_epoch function?

File: ote_cli/ote_cli/utils/hpo.py [hunk 3]
```diff
@@ -297,23 +551,57 @@ def __init__(
...
+                            / train_dataset_size
+                        )
+                elif _is_det_framework_task(task_type) or _is_seg_framework_task(
+                    task_type
+                ):
+                    hpopt_arguments["min_iterations"] = ceil(
+                        env_hp.learning_parameters.learning_rate_warmup_iters
+                        / ceil(train_dataset_size / batch_size)
+                    )
+
```
Comment (eunwoosh) [inline ote_cli/ote_cli/utils/hpo.py:603]: Doesn't code to process the case where "min_iteration" ig bigger than "max_iteration" need to be here?

File: ote_cli/ote_cli/utils/hpo.py [hunk 4]
```diff
@@ -448,6 +733,16 @@ def run(self):
...
+        for key, val in self.deleted_hp.items():
+            best_config[key] = val
+
+        # TODO: is it needed here?
+        # # finetune stage resumes hpo trial, so warmup isn't needed
+        # if task_type == TaskType.DETECTION:
+        #     best_config["learning_parameters.learning_rate_warmup_iters"] = 0
+        # if task_type == TaskType.SEGMENTATION:
+        #     best_config["learning_parameters.learning_rate_fixed_iters"] = 0
+        #     best_config["learning_parameters.learning_rate_warmup_iters"] = 0
```
Comment (eunwoosh) [inline ote_cli/ote_cli/utils/hpo.py:748]: It looks better to move these codes in HpoTrainTask class in get_train_wrapper_task function.
How do you think about?

File: ote_cli/ote_cli/utils/hpo.py [hunk 5]
```diff
@@ -246,44 +466,78 @@ def __init__(
...
-                    "hpo_config.yaml",
-                ),
-                "r",
-                encoding="utf-8",
-            ) as f:
-                hpopt_cfg = yaml.safe_load(f)
-        except FileNotFoundError as err:
-            print("hpo_config.yaml file should be in directory where template.yaml is.")
-            raise err
+        self.deleted_hp = {}
```
Comment (eunwoosh) [inline ote_cli/ote_cli/utils/hpo.py]: I think this variable name is confusing. I'll change this name to decided_hp.

Comment (eunwoosh) [reply ote_cli/ote_cli/utils/hpo.py]: I changed it to fixed_hp and add comments

**Review by goodsong81 — DISMISSED** [review]
LGTM.
Please note that MPA templates would also need similar changes after a while.

**Review by harimkang — DISMISSED** [review]
LGTM

**Review by eunwoosh — CHANGES_REQUESTED** [review]

File: external/mmdetection/detection_tasks/extension/utils/hooks.py (hunk 1 above)
Comment (yunchu) [reply external/mmdetection/detection_tasks/extension/utils/hooks.py:182]: what we want to get for the HPO callback is now iteration number. so we can only put it on this event.

File: ote_cli/ote_cli/utils/hpo.py (hunk 3 above)
Comment (yunchu) [reply ote_cli/ote_cli/utils/hpo.py:603]: we've already put the defense code for that case in the hpopt module. so i don't think it should be here again.

File: ote_cli/ote_cli/utils/hpo.py (hunk 4 above)
Comment (yunchu) [reply ote_cli/ote_cli/utils/hpo.py:748]: those codes are commented out so can you point out again correctly what "these codes" mean?

Comment (eunwoosh) [reply ote_cli/ote_cli/utils/hpo.py:748]: I meant to move code commented out in HpoTrainTask class. But I think it's better to implement this part in each OTE tasks.

**Review by eunwoosh — DISMISSED** [review]

**Review by eunwoosh — APPROVED** [review]

**Review by goodsong81 — APPROVED** [review]

**Pull request merged**